lib/ProductOpener/APITest.pm

```diff
@@ -402,208 +405,229 @@ Note: this setting can be overriden for each test case by specifying a "ua" fiel
-				next;
-			};
-
-			# normalize for comparison
-			if (ref($decoded_json) eq 'HASH') {
-				if (defined $decoded_json->{'products'}) {
-					normalize_products_for_test_comparison($decoded_json->{'products'});
-					if (defined $test_ref->{sort_products_by}) {
-						sort_products_for_test_comparison($decoded_json->{'products'}, $test_ref->{sort_products_by});
-					}
-				}
-				if (defined $decoded_json->{'product'}) {
-					normalize_product_for_test_comparison($decoded_json->{'product'});
+			diag("Response content: " . $response_content);
+			fail($test_case);
+			next;
+		};
+
+		# normalize for comparison
+		if (ref($decoded_json) eq 'HASH') {
+			if (defined $decoded_json->{'products'}) {
+				normalize_products_for_test_comparison($decoded_json->{'products'});
+				if (defined $test_ref->{sort_products_by}) {
+					sort_products_for_test_comparison($decoded_json->{'products'}, $test_ref->{sort_products_by});
 				}
 			}
-
-			is(
-				compare_to_expected_results(
-					$decoded_json, "$expected_result_dir/$test_case.json",
-					$update_expected_results, $test_ref
-				),
-				1,
-				"$test_case - result"
-			);
+			if (defined $decoded_json->{'product'}) {
+				normalize_product_for_test_comparison($decoded_json->{'product'});
+			}
 		}
 
-		# Check if the response content matches what we expect
-		my $must_match = $test_ref->{response_content_must_match};
-		if (    (defined $must_match)
-			and ($response_content !~ /$must_match/i))
-		{
-			fail($test_case);
-			diag("Must match: " . $must_match . "\n" . "Response content: " . $response_content);
-		}
+		is(
+			compare_to_expected_results(
+				$decoded_json, "$expected_result_dir/$test_case.json",
+				$update_expected_results, $test_ref
+			),
+			1,
+			"$test_case - result"
+		);
+	}
 
-		my $must_not_match = $test_ref->{response_content_must_not_match};
-		if (    (defined $must_not_match)
-			and ($response_content =~ /$must_not_match/i))
-		{
-			fail($test_case);
-			diag("Must not match: " . $must_not_match . "\n" . "Response content: " . $response_content);
-		}
+	# Check if the response content matches what we expect
+	my $must_match = $test_ref->{response_content_must_match};
+	if (    (defined $must_match)
+		and ($response_content !~ /$must_match/i))
+	{
+		fail($test_case);
+		diag("Must match: " . $must_match . "\n" . "Response content: " . $response_content);
+	}
+
+	my $must_not_match = $test_ref->{response_content_must_not_match};
+	if (    (defined $must_not_match)
+		and ($response_content =~ /$must_not_match/i))
+	{
+		fail($test_case);
+		diag("Must not match: " . $must_not_match . "\n" . "Response content: " . $response_content);
+	}
+
+}
+
+sub execute_api_tests ($file, $tests_ref, $ua = undef) {
 
+	my ($test_id, $test_dir, $expected_result_dir, $update_expected_results) = (init_expected_results($file));
+
+	$ua = $ua // LWP::UserAgent->new();
+
+	foreach my $test_ref (@$tests_ref) {
+
+		my $response = execute_request($test_ref, $ua);
```

Comment:
You did an incredible job at refactoring. Sorry if it feels picky, but I still have a request: I would check the expected code response even in case of setup, it can help understanding that a test fails because a setup test did go wrong.

Reply:
Good idea. So for setup test cases, I now still call check_requests_response (so that we can check expected status code), but I don't save or compare the actual response body.